File: packages/asdl-slots/src/asdl_slots/lifecycle/free.py
"""Free assigned slots back to the pool."""

from __future__ import annotations

from collections.abc import Sequence

from asdl_slots.context import SlotsCliContext
from asdl_slots.inventory import SlotInventory, build_slot_inventory
from asdl_slots.lifecycle.outcomes import (
    FreedSlot,
    SlotFreeCleanupAction,
    SlotFreeCleanupResult,
    SlotFreeOutcome,
    SlotFreePlan,
    SlotLifecycleFailure,
)
from asdl_slots.lifecycle.release_cleanup import (
    SLOT_RELEASE_ALL_CLEANUP_ACTIONS,
    execute_release_cleanup,
    plan_release_cleanup,
)
from asdl_slots.lifecycle.release_target import (
    ReleaseTargetFailure,
    free_operation_in_progress_message,
    freed_slot_from_record,
    release_assigned_slot_target,
)

SLOT_FREE_ALL_CLEANUP_ACTIONS = SLOT_RELEASE_ALL_CLEANUP_ACTIONS


def plan_free_slots(
    slots_ctx: SlotsCliContext,
    slot_names: Sequence[str],
    *,
    preflight_errors: Sequence[str] = (),
    trunk_branch: str | None = None,
) -> SlotFreePlan | SlotLifecycleFailure:
    """Validate selected slots and return the free plan without mutating state."""
    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )

    state_errors = _validate_assigned_and_clean(slots_ctx, inventory, slot_names)
    all_errors = (*preflight_errors, *state_errors)
    if all_errors:
        return SlotLifecycleFailure(
            error_type="invalid_slot_args",
            message="\n".join(all_errors),
        )

    trunk = trunk_branch if trunk_branch is not None else slots_ctx.git.get_trunk_branch()
    targets: list[FreedSlot] = []
    for slot_name in slot_names:
        record = inventory.find_by_slot(slot_name)
        if record is None or record.branch is None:
            return SlotLifecycleFailure(
                error_type="slot_not_assigned",
                message=(f"{slot_name} is not currently assigned (state changed during planning)."),
            )
        if record.operation is not None:
            return SlotLifecycleFailure(
                error_type="operation_in_progress",
                message=free_operation_in_progress_message(
                    slot_name=record.slot_name,
                    branch_name=record.branch,
                    worktree_path=record.path,
                    operation=record.operation,
                    action="freeing",
                ),
            )
        targets.append(freed_slot_from_record(record))

    return SlotFreePlan(targets=tuple(targets), trunk_branch=trunk)


def execute_free_plan(
    slots_ctx: SlotsCliContext,
    plan: SlotFreePlan,
) -> SlotFreeOutcome | SlotLifecycleFailure:
    """Detach every target in ``plan``, preserving existing recheck semantics."""
    if not plan.targets:
        return SlotFreeOutcome(freed=())

    inventory = build_slot_inventory(
        slots_ctx.git,
        main_repo_root=slots_ctx.repo.main_repo_root,
    )

    freed: list[FreedSlot] = []
    for target in plan.targets:
        result = release_assigned_slot_target(
            slots_ctx.git,
            inventory,
            target,
            plan.trunk_branch,
            operation_action="freeing",
        )
        if isinstance(result, ReleaseTargetFailure):
            return SlotLifecycleFailure(
                error_type=result.error_type,
                message=_partial_failure_message(_free_execution_failure_message(result), freed),
            )
        freed.append(result)

    return SlotFreeOutcome(freed=tuple(freed))


def free_slots(
    slots_ctx: SlotsCliContext,
    slot_names: Sequence[str],
    *,
    preflight_errors: Sequence[str] = (),
    trunk_branch: str | None = None,
) -> SlotFreeOutcome | SlotLifecycleFailure:
    if not slot_names and not preflight_errors:
        return SlotFreeOutcome(freed=())

    plan = plan_free_slots(
        slots_ctx,
        slot_names,
        preflight_errors=preflight_errors,
        trunk_branch=trunk_branch,
    )
    if isinstance(plan, SlotLifecycleFailure):
        return plan
    return execute_free_plan(slots_ctx, plan)


def plan_cleanup_for_free_targets(
    slots_ctx: SlotsCliContext,
    targets: Sequence[FreedSlot],
    cleanup_actions: Sequence[SlotFreeCleanupAction],
    *,
    trunk_branch: str | None = None,
) -> tuple[SlotFreeCleanupResult, ...]:
    """Plan cleanup entries for free targets without mutating PRs or branches."""
    return plan_release_cleanup(
        slots_ctx,
        targets,
        cleanup_actions,
        trunk_branch=trunk_branch,
    )


def execute_cleanup_for_freed_slots(
    slots_ctx: SlotsCliContext,
    freed: Sequence[FreedSlot],
    cleanup_actions: Sequence[SlotFreeCleanupAction],
    *,
    trunk_branch: str | None = None,
) -> tuple[SlotFreeCleanupResult, ...]:
    """Run requested cleanup actions for slots that detached successfully."""
    return execute_release_cleanup(
        slots_ctx,
        freed,
        cleanup_actions,
        trunk_branch=trunk_branch,
    )


def _validate_assigned_and_clean(
    slots_ctx: SlotsCliContext,
    inventory: SlotInventory,
    slot_names: Sequence[str],
) -> tuple[str, ...]:
    errors: list[str] = []
    for slot_name in slot_names:
        record = inventory.find_by_slot(slot_name)
        if record is None or record.branch is None:
            errors.append(
                f"{slot_name} is not currently assigned. Run `slot list` to see the pool."
            )
            continue
        if record.operation is not None:
            errors.append(
                free_operation_in_progress_message(
                    slot_name=record.slot_name,
                    branch_name=record.branch,
                    worktree_path=record.path,
                    operation=record.operation,
                    action="freeing",
                )
            )
            continue
        if slots_ctx.git.has_uncommitted_changes(record.path):
            errors.append(
                f"{slot_name} has uncommitted changes at {record.path}. "
                f"Commit or stash before freeing."
            )
    return tuple(errors)


def _free_execution_failure_message(failure: ReleaseTargetFailure) -> str:
    if failure.reason == "slot_not_assigned":
        return f"{failure.slot_name} is not currently assigned (state changed during free)."
    if failure.reason == "dirty_worktree":
        return (
            f"{failure.slot_name} has uncommitted changes at {failure.worktree_path} "
            "(state changed during free)."
        )
    return failure.message


def _partial_failure_message(base: str, freed: list[FreedSlot]) -> str:
    if not freed:
        return base
    already = ", ".join(f.slot_name for f in freed)
    return f"{base} Already freed: {already}."

File: ts/packages/capabilities/slot/src/lifecycle/free.ts
import type { RepoSlotContext } from "../context.ts";
import { buildSlotInventory, findBySlot, type SlotInventory } from "../inventory.ts";
import { slotOperationMessage, type LifecycleResult } from "./common.ts";
import {
	freedSlotFromRecord,
	releaseAssignedSlotTarget,
	type FreedSlot,
	type ReleaseTargetFailure,
} from "./release-target.ts";

export interface SlotFreePlan {
	targets: readonly FreedSlot[];
	trunkBranch: string;
}

export interface SlotFreeOutcome {
	freed: readonly FreedSlot[];
}

export interface SlotFreeProgressEvent {
	type: "release_started";
	target: FreedSlot;
}

export type SlotFreeProgressReporter = (event: SlotFreeProgressEvent) => void;

export async function planFreeSlots(
	ctx: RepoSlotContext,
	slotNames: readonly string[],
	options: {
		preflightErrors?: readonly string[] | undefined;
		trunkBranch?: string | undefined;
	} = {},
): Promise<LifecycleResult<SlotFreePlan>> {
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	const validation = await validatedFreeTargets(ctx, inventory, slotNames);
	const errors = [...(options.preflightErrors ?? []), ...validation.errors];
	if (errors.length > 0)
		return {
			type: "failure",
			failure: { errorType: "invalid-slot-args", message: errors.join("\n") },
		};
	return {
		type: "ok",
		outcome: {
			targets: validation.targets,
			trunkBranch: options.trunkBranch ?? (await ctx.git.getTrunkBranch()),
		},
	};
}

export async function executeFreePlan(
	ctx: RepoSlotContext,
	plan: SlotFreePlan,
	progress?: SlotFreeProgressReporter | undefined,
): Promise<LifecycleResult<SlotFreeOutcome>> {
	if (plan.targets.length === 0) return { type: "ok", outcome: { freed: [] } };
	const inventory = await buildSlotInventory(ctx.git, { mainRepoRoot: ctx.repo.mainRepoRoot });
	const freed: FreedSlot[] = [];
	for (const target of plan.targets) {
		progress?.({ type: "release_started", target });
		const result = await releaseAssignedSlotTarget({
			git: ctx.git,
			inventory,
			target,
			trunkBranch: plan.trunkBranch,
		});
		if ("reason" in result)
			return {
				type: "failure",
				failure: {
					errorType: result.errorType,
					message: partialFailureMessage(freeExecutionFailureMessage(result), freed),
				},
			};
		freed.push(result);
	}
	return { type: "ok", outcome: { freed } };
}

async function validatedFreeTargets(
	ctx: RepoSlotContext,
	inventory: SlotInventory,
	slotNames: readonly string[],
): Promise<{ targets: readonly FreedSlot[]; errors: readonly string[] }> {
	const errors: string[] = [];
	const targets: FreedSlot[] = [];
	for (const slotName of slotNames) {
		const record = findBySlot(inventory, slotName);
		if (record === null) {
			errors.push(`${slotName} is not currently assigned. Run \`slot list\` to see the pool.`);
			continue;
		}
		if (record.operation !== null) {
			errors.push(slotOperationMessage(record, { action: "freeing" }));
			continue;
		}
		if (record.branch === null) {
			errors.push(`${slotName} is not currently assigned. Run \`slot list\` to see the pool.`);
			continue;
		}
		if (await ctx.git.hasUncommittedChanges(record.path)) {
			errors.push(
				`${slotName} has uncommitted changes at ${record.path}. Commit or stash before freeing.`,
			);
			continue;
		}
		targets.push(freedSlotFromRecord(record));
	}
	return { targets, errors };
}

function freeExecutionFailureMessage(failure: ReleaseTargetFailure): string {
	switch (failure.reason) {
		case "slot-not-assigned":
			return `${failure.slotName} is not currently assigned (state changed during free).`;
		case "operation-in-progress":
			return `${failure.slotName} holds '${failure.branchName}' with a ${failure.operation ?? "operation"} in progress at ${failure.worktreePath}; cannot continue freeing.`;
		case "dirty-worktree":
			return `${failure.slotName} has uncommitted changes at ${failure.worktreePath} (state changed during free).`;
		case "detach-failed":
			return detachFailureMessage(failure);
	}
}

export function detachFailureMessage(failure: ReleaseTargetFailure): string {
	const detachRef = failure.detachRef ?? "target ref";
	const detail = failure.detachError === null ? "" : `: ${failure.detachError}`;
	return `Failed to detach ${failure.slotName} at ${failure.worktreePath} to ${detachRef}${detail}`;
}

function partialFailureMessage(base: string, freed: readonly FreedSlot[]): string {
	if (freed.length === 0) return base;
	return `${base} Already freed: ${freed.map((slot) => slot.slotName).join(", ")}.`;
}
